### PR TITLE
Milestone 1

### PR DESCRIPTION
**Pull request by Arpit1204**
This PR includes the following changes:

- Added the `hotelBooking.cpp` file with the client class definition.
- Added different files like `hotelRoom.txt`, `client.csv` etc.
- The `hotelBooking.cpp` file  demonstrates the use of the `Object and Classes` and `Allocation Memory Dynamically` .

Loom video Links:- [Link 1](https://www.loom.com/share/74c1cf50d3144c2b9cd449b8dfd65bf0?sid=82a451e2-fe90-4310-9bbb-4b08c3c1aa6b), [Link 2](https://www.loom.com/share/b33c03b0e93d4f3cba2c0d27b0e055b7?sid=8399fe2a-2078-4405-8525-af8bbe6a83fd)

### CODE REVIEW

**Review by th3c0d3br34ker — COMMENTED** [review]
hey thanks @Arpit1204, i have reviewed your loom. You have explained everything clearly. 

However, The usage of class is not done properly. Although you have defined a class `client`  to encapsulate data and behaviour related to clients. It has limited OOP design. Also object instantiation is missing.

Overall Comments:
1. Your variable naming conventions are inconsistent. Some variables are named in `camelCase`, while others are in `snake_case`. It's good practice to stick to a consistent naming convention for better code readability.
2. You open the `client.csv` file for both reading and writing in different functions, but you should ensure that the file is closed properly in all code paths, even if an exception occurs.
3. You're using a catch-all `catch (...)` block. It's better to catch specific exceptions if possible, or at least provide a more descriptive error message when an exception occurs.
4. Consider to define Constants for Menu Options.
  Example
  >```
  >const int SEARCH_CLIENT = 1;
  >const int REGISTER_CLIENT = 2;
  >const int EXIT_PROGRAM = 0
  >```
  
  Please address these issues and let me know.

**Comment by Arpit1204** [comment]
Hey thanks @th3c0d3br34ker for the review, I've made most of the requested changes, but could you please provide an example that illustrates the issue mentioned in the second point, where file closure might not be handled correctly?

File: hotelBooking.cpp [hunk 1]
```diff
@@ -39,110 +43,124 @@ class client
...
         {
 
             cout
                 << endl
                 << "Record not found" << endl;
+            clientFile.close();
         }
     }
 
     int create()
```
Comment (th3c0d3br34ker) [inline hotelBooking.cpp]: In the create function, you open the file and write client information to it. You also correctly close the file with `clientFile.close()` before returning the client ID. However, there's a potential issue if an exception is thrown after the file is opened but before it's closed. In this case, the file won't be closed, which could lead to resource leaks and data corruption.

File: hotelBooking.cpp [hunk 2]
```diff
@@ -39,110 +43,124 @@ class client
...
-        client_file << c_id << ","
-                    << c_name << ","
-                    << c_phone_no << ","
-                    << c_address << ","
-                    << c_aadhar << "\n";
+        clientFile << clientId << ","
+                   << clientName << ","
+                   << clientPhoneNo << ","
+                   << clientAddress << ","
+                   << clientAadhar << "\n";
```
Comment (th3c0d3br34ker) [inline hotelBooking.cpp:41]: To handle file closure correctly, you can use a `try` and `catch` block to ensure that the file is closed even if an exception is thrown. Consider to close the file in `finally` blockl

**Review by th3c0d3br34ker — APPROVED** [review]
thanks @Arpit1204, i raised a few concerns regarding your code. this PR is satisfactory for demonstrating classes and objects.

If you are satisfied by your code, you may merge it.

**Comment by Arpit1204** [comment]
Hey @th3c0d3br34ker, thanks for the review I have a doubt, as you told the PR is satisfactory for classes and object concept, but in kalvium.community my marks are 2/5, will they remain the same?

**Comment by th3c0d3br34ker** [comment]
hey @Arpit1204, glad to see that you are looking forward to  improve. 

You may have used classes and object from a development perspective, but not from `OOP` perspective.

In your specific code I could see a `Hotel` class, a `Booking` class, a `FileHandler` class which can be made.

The goal of is concept is to help you break down and abstract away the complexity from `main` method.

As I mentioned, if you are happy with the code you have produced, you may merge it. Refer [here](https://wiki.kalvium.community/doc/oop-review-system-uLd6k98Rca#h-scoring-details-per-concept).

You may either:
1. make changes to your code base and resubmit the assignment (there is always scope to improve at kalvium) or
2. merge this PR and go forward with the next concept.

**Comment by Arpit1204** [comment]
Hey @th3c0d3br34ker, I agree I can make multiple classes and objects and I also plan to do that my next milestones also, but just creating a class and an object and just doing cout to show I did the concept doesn't make sense to me, it's just first milestone, and the concept I covered is classes and Objects, as I move forward I'll be creating more class since there are multiple features needed to be done, 

- Hotel Booking
- Restaurant Booking
- Convention hall booking

These are the classes I plan to implement but as told by Kalvium everything can't be don't in 1st PR and my project can't be just `object and classes`, so I  created a small functional code which is a small part of my project and included the client class and at least applies the concept of `classes and objects`.

The marks are given according to the concept and not according to the project at least for the first milestone, also in `kalvium.community` we need to submit PR according to the concepts. As of `OOP` perspective it isn't the final PR, it's just the starting.

**Comment by th3c0d3br34ker** [comment]
hey @Arpit1204, all the classes which you mentioned appear to be related to `Booking`. The code which you produced also carry marks. 

You also completely missed the following point:

> You may have used classes and object from a development perspective, but not from OOP perspective.

And quoting you:

> "...and at least applies the concept of classes and objects."

which looks like the bare minimum effort to me. I have just evaluated based on your submission. Please consider as I have limited bandwidth. You may book a slot [here](https://calendar.kalvium.com/jainam-desai/30min).

**Review by th3c0d3br34ker — APPROVED** [review]
thanks @Arpit1204 for reverting back after the viva.

as per our discussion in the viva, in the upcoming PRs it is expected that more classes will be implemented and the functionality will be be extended. 

please go ahead and merge this PR.

**Pull request merged**